os/bluestore: asok command to expand block device and Bluestore while OSD is running

JoshuaGabriel opened the pull request:
implements support for expanding BlueStore block devices while the OSD is running. The implementation expands allocators and freelist managers in-place rather than recreating them.

ceph tell osd.X bluestore bluefs-bdev-expand

https://tracker.ceph.com/issues/68797


---

I didn't touch any of the tree allocators because I'm assuming I can just update device_size.
There is also a lot of redundant code that I copied from bluestore::expand_devices that I'm working to cleanup.


the way I was testing this locally, 
```
truncate -s 150G dev/osd0/block
truncate -s 50G dev/osd0/block.db

./bin/ceph tell osd.0 bluestore bluefs-bdev-expand
0 : device size 0x3e800000(1000 MiB) : using 0x1900000(25 MiB)
1 : device size 0x40000000(1 GiB) : using 0xe00000(14 MiB)
2 : device size 0x1900000000(100 GiB) : using 0xd9000(868 KiB) : bluefs used 0x0(0 B) : non-bluefs used 0xd9000(868 KiB)
Expanding DB/WAL...
0 : nothing to do, skipped
1 : Expanding to 0xc80000000(50 GiB)
1 : size updated to 0xc80000000(50 GiB)
2 : Expanding to 0x2580000000(150 GiB)
2 : size updated to 0x2580000000(150 GiB)


then checking outputs of 
./bin/ceph tell osd.0 bluestore bluefs device info
{
    "dev": {
        "device": "BDEV_WAL",
        "total": 1048576000,
        "free": 1022361600,
        "bluefs_used": 26214400
    },
    "dev": {
        "device": "BDEV_DB",
        "total": 53687091200,
        "free": 53672411136,
        "bluefs_used": 14680064
    },
    "dev": {
        "device": "BDEV_SLOW",
        "total": 161061273600,
        "free": 161060212736,
        "bluefs_used": 0,
        "bluefs max available": 161059897344
    }
}

./bin/ceph tell osd.0 bluestore bluefs-bdev-expand
```

separately I would run this
```
./bin/unittest_fastbmap_allocator --gtest_filter="TestAllocatorLevel02.test_expand*"
```


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

JoshuaGabriel committed:
os/bluestore: admin socket for bdev expand

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
blk: update to check new block device size

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: implement expand() for bitmap allocator hierarchy

Implements device expansion support for bitmap-based allocators
BitmapAllocator and HybridAllocator. Bitmap allocators uses allocated
fixed-size bit vectors (L0, L1, L2) and requires vector resizing
and metadata updates when underlying device expands.

Some different cases that are handled:
- padding reclamation when alignment boundaries change
- Sub-granularity expansions (L2 doesn't grow, only L2 bits update)

To be used such as expand() followed by init_add_free() to ensure
available space accounting happens once.

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: expose expand() method for FreelistManager

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
bluestore/bluefs: add expand_device for online device usage

adds support for expanding the device capacity while fs is online

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: add asok for online device expansion

expands bluestore and bluefs while OSD is running.
Instead of recreating allocators, expands allocator and FM
in-place.

ceph tell osd.X bluestore bluefs-bdev-expand

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9160,6 +9160,142 @@ int BlueStore::expand_devices(ostream& out)
+      alloc->init_add_free(size0, aligned_size - size0);
+
+      need_to_destage_allocation_file = true;
+      before_expansion_bdev_size = 0;
```
The assignment is redundant - this var is used during OSD startup phase hence no need to touch it here

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9160,6 +9160,142 @@ int BlueStore::expand_devices(ostream& out)
   return r;
 }
 
+int BlueStore::expand_devices_online(ostream& out)
```
What's about making one more step forward and get rid off the legacy expand_devices() and use the new one from ceph_bluestore_tool as well? Obviouly after proper device mount..

ifed01 commented:
qa/standalone/osd/osd-bluefs-volume-ops.sh has some tests for offline device expansion, mind adding a few tests for online one?

ifed01 commented on src/os/bluestore/BlueFS.cc:
```diff
@@ -915,6 +915,29 @@ void BlueFS::_stop_alloc()
   }
 }
 
+void BlueFS::expand_device(unsigned devid, uint64_t new_size, uint64_t old_size)
```
One should update volume selector (pointed by BlueFS::vselector) as well

JoshuaGabriel committed:
blk/BlockDevice: changes block device size as atomic

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: update bluefs volume selector

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: refactors expand_devices for online and offline use

Refactors device expansion to use a expand_devices()
that handles both online (admin socket) and offline (ceph-bluestore-tool)
use cases.

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
qa/standalone/bluefs: add online device expansion test

verify online device expansion via admin socket works correctly

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel commented:
@ifed01 this should be good to review again

ifed01 commented on qa/standalone/osd/osd-bluefs-volume-ops.sh:
```diff
@@ -492,6 +492,87 @@ function TEST_bluestore_expand() {
     ceph-bluestore-tool --log-file $dir/bluestore_tool.log --path $dir/0 qfsck || return 1
 }
 
+function TEST_bluestore_expand_online() {
```
Additionally it might be helpful to check if bluefs-bdev-expand call works fine if not underlying device expansion took place

JoshuaGabriel committed:
os/bluestore: make Allocator::device_size atomic and protect BitmapAllocator::expand

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
qa/standalone: bdev-expand calls with no underlying device expansion

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: skip bitmapfm exapnd if old and new size are the same

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: remove unused if branch

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: rounds size0 up such that expand is [p2roundup(size0), aligned_size)

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9141,22 +9164,27 @@ int BlueStore::expand_devices(ostream& out)
-      ceph_assert(r == 0);
+      fm->expand(aligned_size, db);
+      alloc->expand(aligned_size);
+      alloc->init_add_free(size0, aligned_size - size0);
```
expand should be this area now [p2roundup(size0), aligned_size)
3451c40812ffa0983712004428460027a4880ad4

JoshuaGabriel committed:
os/bluestore: round old size by min_alloc

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: adds assert for device mounted before expand

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: fix nullptr derefence for bdev in single-device setup

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: open bluestore in read only mode during device expansion

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: dout free space added starts at aligned_old_size

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel committed:
os/bluestore: fix offline device expansion in expand_devices

guard bluefs->expand_device for online-only
fixes offline expansion by reopening db from r/o to r/w
fixes int64_t/uint64_t type mismatch for old_size

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel commented:
@ifed01 added this https://github.com/ceph/ceph/pull/66344/changes/d381a12aff83284d26a6e8f48afb657fcfa4d59d to reopen DB for offline expansion

Let me know if it helps to squash this to a fewer commits

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9036,31 +9036,55 @@ string BlueStore::get_device_path(unsigned id)
+
     auto my_bdev = bluefs->get_block_device(devid);
+    my_bdev->refresh_size();
     uint64_t size = my_bdev ? my_bdev->get_size() : 0;
```
No, bdev can't be null here any more - see lines above. Mind simplifying to size = my_bdev->get_size(); ?

JoshuaGabriel replied:
sorry for the confusion, I added that comment before adding the `    if (!my_bdev) continue;`

although simplified that get_size check in the latest commit https://github.com/ceph/ceph/pull/66344/changes/761900c54ab349ae38a3197122d5e737b2f66aed

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9166,16 +9191,26 @@ int BlueStore::expand_devices(ostream& out)
                 << " : size updated to 0x" << std::hex << size
                 << std::dec << "(" << byte_u_t(size) << ")"
                 << std::endl;
+            // online expand needs to update allocator now
```
A bit confusing comment IMO. Shouldn't that be "... needs to update BlueFS allocator(s) now"?

JoshuaGabriel replied:
fixed the comment https://github.com/ceph/ceph/pull/66344/changes/761900c54ab349ae38a3197122d5e737b2f66aed

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -9219,33 +9254,35 @@ int BlueStore::expand_devices(ostream& out)
-      // Mount in read/write to sync expansion changes
-      // and make sure everything is all right.
-      //
-      before_expansion_bdev_size = size0; // preserve orignal size to permit
```
Likely this line to be brought back and put prior to _open_db_and_around(false) call. It enables allocator's expansion before DB opening in R/W mode which might need that additional space for some internal maintenance(compaction) on startup.
And the following fm->expand & alloc->expand calls to be issued when need_to_close is false only:
if (need_to_close) {
 // do db reopen
} else {
  fm->expand(...);
  alloc->expand(aligned_size);
  uint64_t aligned_size0 = p2roundup(size0, min_alloc_size);
  alloc->init_add_free(aligned_size0, aligned_size - aligned_size0);
}

JoshuaGabriel replied:
that makes sense, should be fixed here https://github.com/ceph/ceph/pull/66344/changes/761900c54ab349ae38a3197122d5e737b2f66aed

the offline expansion should be almost identical to what the code previous besides the refresh_size checks. I wasn't sure in the beginning if that is what we wanted to do, but expanding device before opening in r/w makes the most sense

JoshuaGabriel committed:
os/bluestore: restore previous offline expansion logic

expanding allocator before opening for offline path
fixes to comments and null check

Signed-off-by: Joshua Blanch <joshua.blanch@clyso.com>

JoshuaGabriel commented:
jenkins test make check

JoshuaGabriel commented:
jenkins test api

JoshuaGabriel commented:
jenkins test windows

aclamk commented:
jenkins test windows

Jayaprakash-ibm commented:
RADOS Approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues77604
